torch/csrc/jit/codegen/cuda/expr_simplifier.cpp

```diff
@@ -0,0 +1,430 @@
+
+    if (!changed) {
+      return value;
+    }
+
+    Val* output = IrBuilder::newScalar(*value->getDataType());
+    auto create_fn = def->newObjectFunc();
+    create_fn(
+        def->container(), std::move(new_inputs), {output}, def->attributes());
+    return output;
+  } else {
+    // Handle associative-and-commutative op:
+    // Convert binary ops into flattened op
+    auto output = IrBuilder::newScalar(*value->getDataType());
+    std::vector<Val*> inputs;
+
+    auto append_or_merge_inputs = [&](Val* operand) {
+      auto op = dynamic_cast<FlattenedAssocCommOp*>(operand->definition());
+      if (op != nullptr && op->getOpType() == bop->getBinaryOpType() &&
+          hasSimilarType(op->dtype(), *value->getDataType())) {
+        inputs.insert(inputs.end(), op->inputs().begin(), op->inputs().end());
+        if (op->hasConstantTerm()) {
+          inputs.emplace_back(op->getConstantTerm());
+        }
+      } else {
+        inputs.emplace_back(operand);
+      }
+    };
+
+    append_or_merge_inputs(flatten(bop->lhs()));
+    append_or_merge_inputs(flatten(bop->rhs()));
+
+    IrBuilder::create<FlattenedAssocCommOp>(
+        bop->getBinaryOpType(), output, std::move(inputs));
+    return output;
+  }
+}
+
+// Recursively convert expressions like FlattenedAdd(a, b, c, d) into
+// AddOp(AddOp(AddOp(a, b), c), d))
+Val* unflatten(Val* value, const std::list<ValInfo>& variables) {
+  auto def = value->definition();
+  if (def == nullptr) {
+    return value;
+  }
+  if (isMagicZeroProtectedVariable(value)) {
+    return value;
+  }
+
+  TORCH_INTERNAL_ASSERT(
+      def->outputs().size() == 1,
+      "Expressions with multiple output are not supported");
+
+  auto fop = dynamic_cast<FlattenedAssocCommOp*>(def);
+
+  if (fop == nullptr) {
+    // Handle ops other than flattened op:
+    // Just recursively call `unflatten` on its inputs
+    bool changed = false;
+    std::vector<Val*> new_inputs;
+    new_inputs.reserve(def->inputs().size());
+    for (auto v : def->inputs()) {
+      new_inputs.emplace_back(unflatten(v, variables));
+      if (new_inputs.back() != v) {
+        changed = true;
+      }
+    }
+
+    if (!changed) {
+      return value;
+    }
+
+    Val* output = IrBuilder::newScalar(*value->getDataType());
+    auto create_fn = def->newObjectFunc();
+    create_fn(
+        def->container(), std::move(new_inputs), {output}, def->attributes());
+    return output;
+  } else {
+    // Handle flattened op:
+    // Convert flattened op into original binary ops
+    TORCH_INTERNAL_ASSERT(fop->hasConstantTerm() + fop->inputs().size() >= 2);
+    auto sorted_inputs = fop->sortedInputs(variables);
+    Val* lhs;
+    int64_t next;
+    if (fop->hasConstantTerm()) {
+      lhs = fop->getConstantTerm();
+      next = 0;
+    } else {
+      lhs = unflatten(sorted_inputs.at(0), variables);
+      next = 1;
+    }
+    while (next < sorted_inputs.size()) {
+      auto rhs = unflatten(sorted_inputs.at(next), variables);
```

Comment:
Could you put an example in the comment showing why unflatten on each input is needed. I believe it's because we could have nested flattened expressions from something like:
`addOp(addOp(a, b), mulOp(c, mulOp(d, e))`

Reply:
added